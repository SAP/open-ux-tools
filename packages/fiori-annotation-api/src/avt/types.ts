import type {
    RawAnnotation,
    AnnotationRecord,
    PropertyValue,
    Expression,
    Collection,
    AnnotationList
} from '@sap-ux/vocabularies-types';

export type AVTNode = RawAnnotation | AnnotationRecord | PropertyValue | Expression | Collection | AnnotationList;
