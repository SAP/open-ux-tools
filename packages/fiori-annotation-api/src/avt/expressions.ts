import type {
    AnnotationPathExpression,
    Collection,
    NavigationPropertyPathExpression,
    PathExpression,
    PropertyPathExpression
} from '@sap-ux/vocabularies-types';

/**
 * all attribute or node names that can provide a value representing an Expression (according to typescript definition)
 *  - each entry has the corresponding primitive JS type as string value
 */
export const expressionNames: Record<string, string> = {
    String: 'string',
    Bool: 'boolean',
    Decimal: 'number',
    Date: 'string',
    Float: 'number',
    Int: 'number',
    Path: 'string',
    PropertyPath: 'string',
    AnnotationPath: 'string',
    NavigationPropertyPath: 'string',
    EnumMember: 'string',
    Collection: 'array',
    Record: 'object',
    Apply: 'object',
    Null: 'null'
};

/**
 *
 * @param value - Node.
 * @returns True if node is an expression.
 */
export function isExpression(
    value: Collection[number]
): value is PropertyPathExpression | PathExpression | NavigationPropertyPathExpression | AnnotationPathExpression {
    return typeof value === 'object' && value.type !== undefined && expressionNames[value.type] !== undefined;
}
