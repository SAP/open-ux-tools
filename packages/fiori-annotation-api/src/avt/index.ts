export { convertAnnotationFile, AnnotationListWithOrigins } from './annotations';
export { convertMetadataToAvtSchema } from './metadata';
export {
    convertAnnotationToInternal,
    convertApplyToInternal,
    convertCollectionElement,
    convertCollectionToInternal,
    convertExpressionToInternal,
    convertPrimitiveValueToInternal,
    convertPropertyValueToInternal,
    convertRecordToInternal,
    convertTargetAnnotationsToInternal
} from './to-internal';
export { convertPointerInAnnotationToInternal } from './pointer';
export { resolvePath, isAnnotation, isAnnotationList, isCollection, isRecord } from './utils';
export { AVTNode } from './types';
export { getAvtNodeFromPointer, findAnnotation, findAnnotationByReference } from './find';
export { isExpression } from './expressions';
