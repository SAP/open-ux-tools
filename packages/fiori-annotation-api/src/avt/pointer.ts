import type { Element, ElementName } from '@sap-ux/odata-annotation-core-types';
import { TEXT_TYPE, ELEMENT_TYPE, Edm } from '@sap-ux/odata-annotation-core-types';

import type { JsonPointer } from '../types';
import { expressionNames } from './expressions';

/*
  Fiori Annotation API uses paths (in JsonPointer format) to specify locations inside a (non embedded) annotation value.

  Annotation values are represented as in internal annotation file representation (Element, Attribute, TextNode) and
  paths therein use JsonPointer format.

  Remark: Internal representation of annotation value has more flexibility (choice between attribute and sub element notation)
  Hence - converting path to internal needs to have the actual internal representation.
  Also: error situations can occur in internal representation, which cannot be expressed in AVT format (e.g. multiple values for a PropertyValue)
 */

/**
 * Convert path in annotation to internal representation.
 *
 * @param annotation - Internal representation element.
 * @param pointer - path in external representation
 * @param valueType - Current value type
 * @returns Pointer to an internal representation node.
 */
export function convertPointerInAnnotationToInternal(
    annotation: Element,
    pointer: JsonPointer,
    valueType?: string
): JsonPointer {
    // e.g.
    // "/record/propertyValues/3/value/Collection/0/PropertyPath"
    // into "/content/0/content/3/content/0/content/0/text"
    // based on
    /*
    <Annotation Term="UI.Chart">
        <Record Type="UI.ChartDefinitionType">
            <PropertyValue Property="Title" String="Sales Data"/>
            <PropertyValue Property="Description" String="Sales"/>
            <PropertyValue Property="ChartType" EnumMember="UI.ChartType/Bar"/>
            <PropertyValue Property="Dimensions">
                <Collection>
                    <PropertyPath>Revenue</PropertyPath> // <<<====
    */
    let path: string[] = [''];
    const segments = pointer.split('/').filter((segment) => !!segment);
    let currentElement = annotation;
    let indexedSubElementName: string | undefined; // '' is valid value when looking for all sub element names
    segments.forEach((segment) => {
        if (!currentElement || !path) {
            // further path segments but no element to resolve it: error
            path = [];
            return;
        }
        let result: { subElement?: Element; subElementIndex: number } = { subElementIndex: -1 };
        switch (segment) {
            case 'annotations':
                // n'th Annotation sub element should represent n'th embedded annotation
                indexedSubElementName = Edm.Annotation;
                break;
            case 'record':
            case 'Record':
                // must be record sub element
                result = findSubElement(currentElement, Edm.Record, 0);
                break;
            case 'propertyValues':
                // n'th PropertyValue sub element should represent n'th entry in propertyValues
                indexedSubElementName = Edm.PropertyValue;
                break;
            case 'collection':
            case 'Collection':
                // just goto n'th subElement
                result = findSubElement(currentElement, Edm.Collection, 0);
                indexedSubElementName = '';
                break;
            case 'value':
                // if last segment: stay on current element, if further segments follow: they are handled in default
                path.push(...handleExpressionValueSegment(currentElement, valueType));
                break;
            case 'type':
                // type attribute in a Record
                path.push('attributes', 'Type', 'value');
                break;
            default:
                if (Number.isNaN(Number(segment))) {
                    // segment is not an index
                    const { subElement, clearPath, pathSegments } = handleStringSegment(segment, currentElement);
                    result.subElement = subElement;
                    path = clearPath ? [] : [...path, ...(pathSegments ?? [])];
                } else if (indexedSubElementName !== null) {
                    // segment is an index - find sub element with specific name
                    result = findSubElement(currentElement, indexedSubElementName, Number(segment));
                    indexedSubElementName = undefined;
                }
                break;
        }
        if (result?.subElementIndex >= 0) {
            path.push('content', result.subElementIndex.toString());
        }
        if (result?.subElement) {
            currentElement = result.subElement;
        }
    });
    return path.join('/');
}

/**
 *
 * @param currentElement
 * @param valueType
 */
function handleExpressionValueSegment(currentElement: Element, valueType?: string): string[] {
    if (valueType === undefined) {
        return [];
    }
    if (currentElement.attributes[valueType]) {
        return ['attributes', valueType];
    }
    const elementIndex = currentElement.content.findIndex(
        (element) => element.type === ELEMENT_TYPE && element.name === valueType
    );
    if (elementIndex !== -1) {
        return ['content', elementIndex.toString()];
    }
    return [];
}

/**
 *
 * @param segment
 * @param currentElement
 */
function handleStringSegment(
    segment: string,
    currentElement: Element
): { pathSegments?: string[]; subElement?: Element; clearPath?: boolean } {
    if (expressionNames[segment]) {
        // primitive type name; represented either as attribute or text sub node; should be the last segment
        return handleExpressionNameSegment(currentElement, segment);
    } else if (currentElement.attributes[firstCharToUpper(segment)]) {
        // can be qualifier, term, type: all are EDMX attribute names
        return { pathSegments: ['attributes', firstCharToUpper(segment), 'value'] };
    } else if (segment === 'name') {
        // propertyValue.name is value of EDMX attribute 'Property'
        return { pathSegments: ['attributes', Edm.Property, 'value'] };
    } else {
        return { clearPath: true };
    }
}

/**
 *
 * @param currentElement
 * @param segment
 */
function handleExpressionNameSegment(
    currentElement: Element,
    segment: string
): { pathSegments?: string[]; subElement?: Element; clearPath?: boolean } {
    if (currentElement?.attributes[segment]) {
        // primitive type name represented as attribute
        return { pathSegments: ['attributes', segment, 'value'] };
    }
    if (currentElement.name === segment) {
        // point to first text node
        const firstTextNodeIndex = getFirstTextNodeIndex(currentElement);
        if (firstTextNodeIndex < 0) {
            return { clearPath: true };
        } else {
            return { pathSegments: ['content', firstTextNodeIndex.toString()] };
        }
    }
    // find sub node representing primitive value and point to it's first text node
    const { subElement, subElementIndex } = findSubElement(currentElement, segment, 0);
    if (subElementIndex < 0) {
        return { clearPath: true, subElement };
    }
    // subElementIndex found
    const pathSegments: string[] = ['content', subElementIndex.toString()];
    const firstTextNodeIndex = getFirstTextNodeIndex(currentElement.content[subElementIndex] as Element);
    if (firstTextNodeIndex < 0) {
        return { clearPath: true, subElement };
    }
    pathSegments.push('content', firstTextNodeIndex.toString());
    return { subElement, pathSegments };
}

/**
 *
 * @param element
 */
function getFirstTextNodeIndex(element: Element): number {
    return element.content.findIndex((subNode) => subNode.type === TEXT_TYPE);
}

/**
 *
 * @param element
 * @param subElementName
 * @param occurrence
 */
function findSubElement(
    element: Element,
    subElementName: ElementName | undefined,
    occurrence: number
): { subElement?: Element; subElementIndex: number } {
    const result: { subElement?: Element; subElementIndex: number } = {
        subElementIndex: -1
    };
    let currentMatchIndex = -1;
    for (let index = 0; index < element.content.length; index++) {
        const child = element.content[index];
        if (child.type === ELEMENT_TYPE && (!subElementName || child.name === subElementName)) {
            currentMatchIndex++;
            if (currentMatchIndex === occurrence) {
                result.subElement = child;
                result.subElementIndex = index;
            }
        }
    }
    return result;
}

/**
 *
 * @param source
 */
function firstCharToUpper(source: string): string {
    let out: string = source;
    if (source) {
        out = source.slice(0, 1).toUpperCase() + source.slice(1);
    }
    return out;
}
