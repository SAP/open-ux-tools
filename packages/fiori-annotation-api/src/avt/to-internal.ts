import type {
    AnnotationRecord,
    Apply,
    Collection,
    Expression,
    PropertyValue,
    RawAnnotation
} from '@sap-ux/vocabularies-types';
import { resolveName, toAliasQualifiedName } from '@sap-ux/odata-annotation-core';
import type { AliasInformation, Element, ElementChild, PathValue, Target } from '@sap-ux/odata-annotation-core-types';
import {
    Edm,
    createElementNode,
    createAttributeNode,
    createTextNode,
    ELEMENT_TYPE,
    TEXT_TYPE
} from '@sap-ux/odata-annotation-core-types';
import { expressionNames, isExpression } from './expressions';
import { getAliasedEnumMember, resolveEnumMemberValue, resolvePath } from './utils';
import type { AnnotationListWithOrigins } from './annotations';

/**
 * Convert target annotations to internal format.
 *
 * @param targetAnnotations - AVT annotation list
 * @param aliasInfo - Alias Information.
 * @returns Internal representation of the Target.
 */
export function convertTargetAnnotationsToInternal(
    targetAnnotations: AnnotationListWithOrigins,
    aliasInfo: AliasInformation
): Target {
    const target = buildEmptyTarget(getAliasedPath(aliasInfo, targetAnnotations.target));
    target.terms = targetAnnotations.annotations.map((annotation) =>
        convertAnnotationToInternal(annotation, aliasInfo)
    );
    return target;
}

/**
 * Build empty target (generic annotation file format).
 *
 * @param path - Target name.
 * @returns Internal representation of the target.
 */
export function buildEmptyTarget(path: string): Target {
    return { type: 'target', name: path, terms: [], range: undefined, nameRange: undefined, termsRange: undefined };
}

/**
 * Convert annotation to internal format.
 *
 * @param annotation - Annotation.
 * @param aliasInfo - Alias Information.
 * @returns Internal representation of the annotation.
 */
export function convertAnnotationToInternal(annotation: RawAnnotation, aliasInfo: AliasInformation): Element {
    const annotationElement = createElementNode({ name: Edm.Annotation });
    annotationElement.attributes[Edm.Term] = createAttributeNode(
        Edm.Term,
        toAliasQualifiedName(annotation.term, aliasInfo)
    );
    if (annotation.qualifier) {
        annotationElement.attributes[Edm.Qualifier] = createAttributeNode(Edm.Qualifier, annotation.qualifier);
    }
    if (annotation.collection) {
        annotationElement.content = annotationElement.content || [];
        annotationElement.content.push(convertCollectionToInternal(aliasInfo, annotation.collection));
    } else if (annotation.record) {
        annotationElement.content = annotationElement.content || [];
        annotationElement.content.push(convertRecordToInternal(aliasInfo, annotation.record));
    } else if (annotation.value) {
        convertExpressionToInternal(aliasInfo, annotation.value, annotationElement);
    }
    if (annotation.annotations?.length) {
        // add embedded annotations
        annotation.annotations.forEach((annotation) => {
            annotationElement.content = annotationElement.content || [];
            annotationElement.content.push(convertAnnotationToInternal(annotation, aliasInfo));
        });
    }
    return annotationElement;
}

/**
 * Convert collection to internal format.
 *
 * @param aliasInfo - Alias Information.
 * @param collection - Collection.
 * @returns Internal representation of the collection.
 */
export function convertCollectionToInternal(aliasInfo: AliasInformation, collection: Collection): Element {
    const collectionElement: Element = createElementNode({ name: Edm.Collection });
    for (const entry of collection) {
        const entryNode = convertCollectionElement(aliasInfo, entry);
        if (entryNode) {
            collectionElement.content.push(entryNode);
        }
    }
    return collectionElement;
}

/**
 *
 * @param aliasInfo - Alias Information.
 * @param entry - Collection entry.
 * @returns Internal representation of the collection entry.
 */
export function convertCollectionElement(
    aliasInfo: AliasInformation,
    entry: Collection[number]
): ElementChild | undefined {
    if (typeof entry === 'object') {
        if (isExpression(entry)) {
            // entry is expression
            return convertExpressionToInternal(aliasInfo, entry);
        } else {
            // entry must be record
            return convertRecordToInternal(aliasInfo, entry);
        }
    } else if (typeof entry === 'string') {
        return createElementNode({ name: Edm.String, content: [createTextNode(entry.toString())] });
    } else if (typeof entry === 'boolean') {
        // obvious extension of annotation.record definition
        return createElementNode({ name: Edm.Bool, content: [createTextNode(entry ? 'true' : 'false')] });
    }
    return undefined;
}

/**
 * Convert record to internal format.
 *
 * @param aliasInfo - Alias Information.
 * @param record - Record.
 * @returns Internal representation of the record.
 */
export function convertRecordToInternal(aliasInfo: AliasInformation, record: AnnotationRecord): Element {
    const recordElement: Element = createElementNode({ name: Edm.Record });
    if (record.type) {
        recordElement.attributes[Edm.Type] = createAttributeNode(
            Edm.Type,
            toAliasQualifiedName(record.type, aliasInfo)
        );
    }
    record.propertyValues.forEach((propertyValue) => {
        const propValueElement = convertPropertyValueToInternal(aliasInfo, propertyValue);
        recordElement.content.push(propValueElement);
    });
    if (record.annotations?.length) {
        // add embedded annotations
        recordElement.content = recordElement.content || [];
        record.annotations.forEach((annotation: RawAnnotation) => {
            recordElement.content.push(convertAnnotationToInternal(annotation, aliasInfo));
        });
    }
    return recordElement;
}

/**
 *
 * @param aliasInfo - Alias Information.
 * @param propertyValue - Property value.
 * @returns Internal representation of the property value.
 */
export function convertPropertyValueToInternal(aliasInfo: AliasInformation, propertyValue: PropertyValue): Element {
    const propValueElement: Element = createElementNode({ name: Edm.PropertyValue });
    propValueElement.attributes[Edm.Property] = createAttributeNode(Edm.Property, propertyValue.name);
    convertExpressionToInternal(aliasInfo, propertyValue.value, propValueElement);
    if (propertyValue.annotations?.length) {
        // add embedded annotations
        propValueElement.content = propValueElement.content || [];
        propertyValue.annotations.forEach((annotation: RawAnnotation) => {
            propValueElement.content.push(convertAnnotationToInternal(annotation, aliasInfo));
        });
    }
    return propValueElement;
}

/**
 * Convert expression to internal representation.
 *
 * @param aliasInfo - Alias Information.
 * @param value - Expression value.
 * @param hostElement add value to this element as host (e.g. add for elements Annotation and PropertyValue)
 * @returns Internal representation of the expression.
 */
export function convertExpressionToInternal(
    aliasInfo: AliasInformation,
    value: Expression,
    hostElement?: Element
): Element | undefined {
    const elementName = expressionNames[value.type] ? value.type : '';
    let element = hostElement ?? (elementName ? createElementNode({ name: elementName }) : undefined);
    if (!element) {
        return undefined;
    }
    let collectionElement: Element;
    let recordElement: Element;
    let nullElement: Element;
    let primitiveValue: string;
    switch (value.type) {
        case 'Collection':
            collectionElement = convertCollectionToInternal(aliasInfo, value.Collection);
            element = consumeElement(element, collectionElement, hostElement);
            break;
        case 'Record':
            recordElement = convertRecordToInternal(aliasInfo, value.Record);
            element = consumeElement(element, recordElement, hostElement);
            break;
        case 'Apply':
            recordElement = convertApplyToInternal(aliasInfo, value.Apply);
            element = consumeElement(element, recordElement, hostElement);
            break;
        case 'Null':
            nullElement = createElementNode({ name: elementName });
            element = consumeElement(element, nullElement, hostElement);
            break;
        case 'Unknown':
            return undefined;
        default: {
            // value type is EDMX primitive expression
            const rawPrimitiveValue = (value as any)[value.type]; // There is always a property with on the object as type name, Typescript does not infer this case as expected
            primitiveValue = convertPrimitiveValueToInternal(value.type, rawPrimitiveValue, aliasInfo);
            if (hostElement) {
                // add value to host element as attribute
                element.attributes = element.attributes || {};
                element.attributes[value.type] = createAttributeNode(value.type, primitiveValue);
            } else {
                // add value as single text sub node
                element.content.push(createTextNode(primitiveValue));
            }
            break;
        }
    }
    return element;
}

/**
 *
 * @param aliasInfo - Alias Information.
 * @param apply - Apply expression.
 * @returns Internal representation of apply.
 */
export function convertApplyToInternal(aliasInfo: AliasInformation, apply: Apply): Element {
    // Apply value is regular internal representation (without alias)
    const clone: Element = JSON.parse(JSON.stringify(apply));
    return replaceAliasInElement(clone, aliasInfo, true);
}

/**
 *
 * @param type - Type of primitive value.
 * @param value - Primitive value.
 * @param aliasInfo - Alias Information.
 * @returns Internal representation of primitive value
 */
export function convertPrimitiveValueToInternal(
    type: string,
    value: string | number | boolean,
    aliasInfo: AliasInformation
): string {
    const text = value.toString();
    if (!text) {
        return text;
    } else if (type === Edm.EnumMember) {
        return getAliasedEnumMember(aliasInfo, text);
    } else if (type.indexOf('Path') >= 0) {
        return getAliasedPath(aliasInfo, text);
    } else if (type === Edm.Type || type === Edm.Term) {
        return toAliasQualifiedName(text, aliasInfo);
    } else {
        return text;
    }
}

function consumeElement(element: Element, collectionElement: Element, hostElement?: Element): Element {
    if (hostElement) {
        element.content = element.content || [];
        element.content.push(collectionElement); // add whole collection as sub element
    } else {
        element = collectionElement;
    }
    return element;
}

function replaceAliasInElement(element: Element, aliasInfo: AliasInformation, reverse?: boolean): Element {
    // replace aliases in all attributes/sub nodes with full namespaces (reverse = true ? vice versa):
    const result = element;
    removeEmptyTextNodes(result);
    replaceAliasInAttributes(result, aliasInfo, reverse);
    replaceAliasInSubNodes(result, aliasInfo, reverse);
    return result;
}

function replaceAliasInAttributes(result: Element, aliasInfo: AliasInformation, reverse?: boolean): void {
    // in attributes: term or type attributes, enumValue and any path values provided as attributes
    Object.keys(result.attributes || {}).forEach((attributeName) => {
        const attribute = result.attributes[attributeName];
        switch (attributeName) {
            case Edm.Term:
            case Edm.Type:
                attribute.value = reverse
                    ? toAliasQualifiedName(attribute.value, aliasInfo)
                    : resolveName(attribute.value, aliasInfo.aliasMap)?.qName;
                break;
            case Edm.EnumMember:
                attribute.value = reverse
                    ? getAliasedEnumMember(aliasInfo, attribute.value)
                    : resolveEnumMemberValue(aliasInfo.aliasMap, aliasInfo.currentFileNamespace, attribute.value);
                break;
            default:
                if (attributeName.endsWith('Path')) {
                    attribute.value = reverse
                        ? getAliasedPath(aliasInfo, attribute.value)
                        : resolvePath(aliasInfo.aliasMap, aliasInfo.currentFileNamespace, attribute.value);
                }
                break;
        }
    });
}

function replaceAliasInSubNodes(result: Element, aliasInfo: AliasInformation, reverse?: boolean): void {
    for (const subNode of result.content) {
        if (subNode.type === ELEMENT_TYPE) {
            replaceAliasInElement(subNode, aliasInfo, reverse);
        } else if (subNode.type === TEXT_TYPE && result.name === Edm.EnumMember) {
            subNode.text = reverse
                ? getAliasedEnumMember(aliasInfo, subNode.text)
                : resolveEnumMemberValue(aliasInfo.aliasMap, aliasInfo.currentFileNamespace, subNode.text);
        } else if (subNode.type === TEXT_TYPE && result.name.endsWith('Path')) {
            subNode.text = reverse
                ? getAliasedPath(aliasInfo, subNode.text)
                : resolvePath(aliasInfo.aliasMap, aliasInfo.currentFileNamespace, subNode.text);
        }
    }
}

function removeEmptyTextNodes(result: Element): void {
    if ((result.content ?? []).some((entry) => entry.type === ELEMENT_TYPE)) {
        // sub elements present: filter out empty text nodes
        result.content = result.content.filter((entry) => !(entry.type === TEXT_TYPE && !(entry.text ?? '').trim()));
    }
}

function getAliasedSegment(aliasInfo: AliasInformation, segment: string): string {
    const [path, term] = segment.split('@');
    if (term) {
        return `${path}@${toAliasQualifiedName(term, aliasInfo)}`;
    } else if (segment.indexOf('.') > -1) {
        return toAliasQualifiedName(segment, aliasInfo);
    } else {
        return segment;
    }
}

function getAliasedPath(aliasInfo: AliasInformation, path: PathValue): PathValue {
    return path
        .split('/')
        .map((segment) => getAliasedSegment(aliasInfo, segment))
        .join('/');
}
