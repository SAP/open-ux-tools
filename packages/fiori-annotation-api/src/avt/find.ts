import type { RawAnnotation, AnnotationList } from '@sap-ux/vocabularies-types';

import { Edm } from '@sap-ux/odata-annotation-core-types';
import type { AliasInformation, AnnotationFile, Element, Target } from '@sap-ux/odata-annotation-core-types';

import type { AnnotationReference } from '../types';
import { getElementAttributeValue, parsePath, resolveName, toFullyQualifiedPath } from '@sap-ux/odata-annotation-core';
import { ApiErrorCode, ApiError } from '../error';
import { annotationReferenceToString } from '../utils';
import type { NamespaceMap } from './utils';
import type { AVTNode } from './types';
import { convertPointerInAnnotationToInternal } from './pointer';

/**
 * Finds a node matching to the pointer from root.
 *
 * @param root - Root node.
 * @param pointer - Pointer to a node.
 * @returns Node matching pointer.
 */
export function getAvtNodeFromPointer(root: RawAnnotation, pointer: string): AVTNode | undefined {
    const segments = pointer.split('/');
    if (segments[0] === '') {
        segments.shift(); // Remove the first empty segment
    }
    if (segments.length === 0) {
        return undefined;
    }
    let node: AVTNode = root;
    for (const segment of segments) {
        const next: AVTNode | undefined = (node as unknown as { [key: string]: AVTNode })[segment];
        if (next) {
            node = next;
        } else {
            return undefined;
        }
    }
    return node;
}

/**
 *  Fins first annotation matching the reference.
 *
 * @param annotationLists - Annotation lists.
 * @param reference - Annotation reference.
 * @returns Matching annotation.
 */
export function findAnnotation(
    annotationLists: AnnotationList[],
    reference: AnnotationReference
): RawAnnotation | undefined {
    for (const annotationList of annotationLists) {
        if (annotationList.target === reference.target) {
            for (const annotation of annotationList.annotations) {
                if (annotation.term === reference.term && annotation.qualifier === reference.qualifier) {
                    return annotation;
                }
            }
        }
    }
    return undefined;
}

export interface SearchResult {
    element: Element;
    target: Target;
    targetPointer: string;
    internalPointer: string;
}

/**
 *
 * @param aliasInfo - Alias information.
 * @param file - Internal representation.
 * @param mergeMap - Annotation merge map.
 * @param reference - Annotation reference.
 * @param pointer - Pointer to a node.
 * @param mergeAnnotations - Flag indicating if annotations were merged.
 * @param valueType - Current value type.
 * @returns Annotation search result.
 */
export function findAnnotationByReference(
    aliasInfo: AliasInformation,
    file: AnnotationFile,
    mergeMap: Record<string, string>,
    reference: AnnotationReference,
    pointer: string,
    mergeAnnotations: boolean,
    valueType?: string
): SearchResult {
    // find generic representation of annotation (take first match for term and qualifier)
    for (let i = 0; i < file.targets.length; i++) {
        const target = file.targets[i];
        const targetName = resolvePath(aliasInfo.aliasMap, aliasInfo.currentFileNamespace, target.name);
        if (targetName !== reference.target) {
            continue;
        }
        for (let j = 0; j < target.terms.length; j++) {
            const term = target.terms[j];
            const qualifier = getElementAttributeValue(term, Edm.Qualifier);
            const termName = resolveName(getElementAttributeValue(term, Edm.Term), aliasInfo.aliasMap).qName;

            if (
                reference.term === termName &&
                ((!qualifier && !reference.qualifier) || qualifier === reference.qualifier)
            ) {
                const q = qualifier ? '#' + qualifier : '';
                const id = `${targetName}@${termName}${q}`;
                if (mergeAnnotations) {
                    const mergedSegments = pointer.split('/').filter((x) => !!x);
                    const keyRoot = `${id}/0/0`;
                    for (let segmentIndex = mergedSegments.length; segmentIndex > 0; segmentIndex--) {
                        const key = `${keyRoot}/${mergedSegments.slice(0, segmentIndex).join('/')}`;
                        const match = mergeMap[key];
                        if (match) {
                            const [, targetIndexSegment, termIndexSegment, ...mappedSourceSegments] = match
                                .replace(id, '')
                                .split('/');
                            const targetIndex = parseInt(targetIndexSegment, 10);
                            const termIndex = parseInt(termIndexSegment, 10);
                            const sourceTerm = file.targets[targetIndex].terms[termIndex];
                            const internalPointer = convertPointerInAnnotationToInternal(
                                sourceTerm,
                                '/' + [...mappedSourceSegments, ...mergedSegments.slice(segmentIndex)].join('/'),
                                valueType
                            );
                            return {
                                element: sourceTerm,
                                target: file.targets[targetIndex],
                                targetPointer: `/targets/${targetIndex}/terms/${termIndex}`,
                                internalPointer
                            };
                        }
                    }
                }
                const internalPointer = convertPointerInAnnotationToInternal(term, pointer, valueType);
                return { element: term, target, targetPointer: `/targets/${i}/terms/${j}`, internalPointer };
            }
        }
    }
    const path = annotationReferenceToString(reference);
    throw new ApiError(`Term '${path}' does not exist`, ApiErrorCode.General);
}

function resolvePath(namespaceMap: NamespaceMap, currentNamespace: string, path: string): string {
    const parsedPath = parsePath(path);
    parsedPath.segments = parsedPath.segments.map((segment, i) => {
        if (segment.namespaceOrAlias === undefined && i === 0) {
            return {
                ...segment,
                namespaceOrAlias: currentNamespace
            };
        }
        return segment;
    });
    return toFullyQualifiedPath(namespaceMap, currentNamespace, parsedPath);
}
