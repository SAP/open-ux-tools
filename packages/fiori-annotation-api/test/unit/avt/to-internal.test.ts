import type { Expression, StringExpression } from '@sap-ux/vocabularies-types';
import { createElementNode, type AliasInformation } from '@sap-ux/odata-annotation-core-types';

import { convertExpressionToInternal } from '../../../src/avt/to-internal';

describe('avt to internal', () => {
    const aliasInfo: AliasInformation = {
        aliasMap: { UI: 'com.sap.vocabularies.UI.v1', 'com.sap.vocabularies.UI.v1': 'com.sap.vocabularies.UI.v1' },
        aliasMapMetadata: {},
        aliasMapVocabulary: {},
        currentFileNamespace: '',
        reverseAliasMap: { 'com.sap.vocabularies.UI.v1': 'UI' }
    };

    describe('convertExpressionToInternal', () => {
        function testConversion(expression: Expression) {
            expect(convertExpressionToInternal(aliasInfo, expression)).toMatchSnapshot();
            expect(
                convertExpressionToInternal(aliasInfo, expression, createElementNode({ name: 'Annotation' }))
            ).toMatchSnapshot();
        }
        test('Collection', () => {
            testConversion({ type: 'Collection', Collection: [] });
        });
        test('Collection with string values', () => {
            testConversion({ type: 'Collection', Collection: ['test'] as unknown as StringExpression[] });
        });
        test('Record', () => {
            testConversion({
                type: 'Record',
                Record: { propertyValues: [], type: 'com.sap.vocabularies.UI.v1.DataField' }
            });
        });
        // Currently not supported
        test('If', () => {
            testConversion({ type: 'If', $If: [{ type: 'Null' }, { type: 'Null' }, { type: 'Null' }] });
        });
        // Currently not supported
        test('Apply', () => {
            testConversion({
                type: 'Apply',
                $Apply: [
                    '#Supplier-displayFactSheet?Supplier={SUP}',
                    {
                        type: 'LabeledElement',
                        $Name: 'RSP',
                        $LabeledElement: {
                            type: 'Path',
                            $Path: 'SupplierEvalResponse'
                        }
                    }
                ],
                $Function: 'odata.fillUriTemplate'
            });
        });
        test('Null', () => {
            testConversion({ type: 'Null' });
        });
        test('Unknown', () => {
            testConversion({ type: 'Unknown' });
        });
        test('Path', () => {
            testConversion({ type: 'Path', Path: 'test_path' });
        });
    });
});
