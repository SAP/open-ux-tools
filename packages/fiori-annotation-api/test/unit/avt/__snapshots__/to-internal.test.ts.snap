// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`avt to internal convertExpressionToInternal Collection 1`] = `
Object {
  "attributes": Object {},
  "content": Array [],
  "name": "Collection",
  "type": "element",
}
`;

exports[`avt to internal convertExpressionToInternal Collection 2`] = `
Object {
  "attributes": Object {},
  "content": Array [
    Object {
      "attributes": Object {},
      "content": Array [],
      "name": "Collection",
      "type": "element",
    },
  ],
  "name": "Annotation",
  "type": "element",
}
`;

exports[`avt to internal convertExpressionToInternal If 1`] = `undefined`;

exports[`avt to internal convertExpressionToInternal If 2`] = `
Object {
  "attributes": Object {
    "If": Object {
      "name": "If",
      "type": "attribute",
      "value": "[object Object]",
    },
  },
  "content": Array [],
  "name": "Annotation",
  "type": "element",
}
`;

exports[`avt to internal convertExpressionToInternal Null 1`] = `
Object {
  "attributes": Object {},
  "content": Array [],
  "name": "Null",
  "type": "element",
}
`;

exports[`avt to internal convertExpressionToInternal Null 2`] = `
Object {
  "attributes": Object {},
  "content": Array [
    Object {
      "attributes": Object {},
      "content": Array [],
      "name": "Null",
      "type": "element",
    },
  ],
  "name": "Annotation",
  "type": "element",
}
`;

exports[`avt to internal convertExpressionToInternal Path 1`] = `
Object {
  "attributes": Object {},
  "content": Array [
    Object {
      "text": "test_path",
      "type": "text",
    },
  ],
  "name": "Path",
  "type": "element",
}
`;

exports[`avt to internal convertExpressionToInternal Path 2`] = `
Object {
  "attributes": Object {
    "Path": Object {
      "name": "Path",
      "type": "attribute",
      "value": "test_path",
    },
  },
  "content": Array [],
  "name": "Annotation",
  "type": "element",
}
`;

exports[`avt to internal convertExpressionToInternal Record 1`] = `
Object {
  "attributes": Object {
    "Type": Object {
      "name": "Type",
      "type": "attribute",
      "value": "UI.DataField",
    },
  },
  "content": Array [],
  "name": "Record",
  "type": "element",
}
`;

exports[`avt to internal convertExpressionToInternal Record 2`] = `
Object {
  "attributes": Object {},
  "content": Array [
    Object {
      "attributes": Object {
        "Type": Object {
          "name": "Type",
          "type": "attribute",
          "value": "UI.DataField",
        },
      },
      "content": Array [],
      "name": "Record",
      "type": "element",
    },
  ],
  "name": "Annotation",
  "type": "element",
}
`;

exports[`avt to internal convertExpressionToInternal Unknown 1`] = `undefined`;

exports[`avt to internal convertExpressionToInternal Unknown 2`] = `undefined`;
